In [25]:
import pandas as pd

In [26]:
data = pd.read_csv('/workspaces/machine-learning-python-template/data/raw/AB_NYC_2019.csv')
data.head()

id                                              name  host_id  \
0  2539                Clean & quiet apt home by the park     2787   
1  2595                             Skylit Midtown Castle     2845   
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632   
3  3831                   Cozy Entire Floor of Brownstone     4869   
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192   

     host_name neighbourhood_group neighbourhood  latitude  longitude  \
0         John            Brooklyn    Kensington  40.64749  -73.97237   
1     Jennifer           Manhattan       Midtown  40.75362  -73.98377   
2    Elisabeth           Manhattan        Harlem  40.80902  -73.94190   
3  LisaRoxanne            Brooklyn  Clinton Hill  40.68514  -73.95976   
4        Laura           Manhattan   East Harlem  40.79851  -73.94399   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room    149               1                  9  2018-10-19   
1  Entire home/apt    225               1                 45  2019-05-21   
2     Private room    150               3                  0         NaN   
3  Entire home/apt     89               1                270  2019-07-05   
4  Entire home/apt     80              10                  9  2018-11-19   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2                NaN                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0

In [27]:
data.shape

(48895, 16)

In [28]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48895 entries, 0 to 48894
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              48895 non-null  int64  
 1   name                            48879 non-null  object 
 2   host_id                         48895 non-null  int64  
 3   host_name                       48874 non-null  object 
 4   neighbourhood_group             48895 non-null  object 
 5   neighbourhood                   48895 non-null  object 
 6   latitude                        48895 non-null  float64
 7   longitude                       48895 non-null  float64
 8   room_type                       48895 non-null  object 
 9   price                           48895 non-null  int64  
 10  minimum_nights                  48895 non-null  int64  
 11  number_of_reviews               48895 non-null  int64  
 12  last_review                     

In [29]:
data.nunique()

id                                48895
name                              47905
host_id                           37457
host_name                         11452
neighbourhood_group                   5
neighbourhood                       221
latitude                          19048
longitude                         14718
room_type                             3
price                               674
minimum_nights                      109
number_of_reviews                   394
last_review                        1764
reviews_per_month                   937
calculated_host_listings_count       47
availability_365                    366
dtype: int64

- Eliminar información irrelevante

En principio vemos que las columna id, host_id se pueden eliminar

host_name tambien se puede eliminar, ya que de entre todos los anuncios listados (48895), 48874 son distintos, es decir, solo hay 21 inmuebles que podrían tener el mismo propietario, 0,04% de la muestra. Por tanto, podemos predecir ya que la identidad del propietario no tiene mayor impacto en la cuestion.

En cuanto a la columna 'name', que vemos que contiene una pequeña descripción, nos encontramos en una situacion parecida, sin embargo,
puesto que se trata de un pequeño 'titular', podemos pensar que no tiene relacion con el precio, pero si es posible que este titular tenga
impacto sobre la demanda, lo que tiene impacto en el precio. 

In [30]:
data[['name', 'number_of_reviews', 'price', 'reviews_per_month', 'calculated_host_listings_count']]

name  number_of_reviews  \
0                     Clean & quiet apt home by the park                  9   
1                                  Skylit Midtown Castle                 45   
2                    THE VILLAGE OF HARLEM....NEW YORK !                  0   
3                        Cozy Entire Floor of Brownstone                270   
4       Entire Apt: Spacious Studio/Loft by central park                  9   
...                                                  ...                ...   
48890    Charming one bedroom - newly renovated rowhouse                  0   
48891      Affordable room in Bushwick/East Williamsburg                  0   
48892            Sunny Studio at Historical Neighborhood                  0   
48893               43rd St. Time Square-cozy single bed                  0   
48894  Trendy duplex in the very heart of Hell's Kitchen                  0   

       price  reviews_per_month  calculated_host_listings_count  
0        149               0.21                               6  
1        225               0.38                               2  
2        150                NaN                               1  
3         89               4.64                               1  
4         80               0.10                               1  
...      ...                ...                             ...  
48890     70                NaN                               2  
48891     40                NaN                               2  
48892    115                NaN                               1  
48893     55                NaN                               6  
48894     90                NaN                               1  

[48895 rows x 5 columns]

In [31]:
data[['number_of_reviews', 'reviews_per_month', 'price']].corr()

number_of_reviews  reviews_per_month     price
number_of_reviews           1.000000           0.549868 -0.047954
reviews_per_month           0.549868           1.000000 -0.030608
price                      -0.047954          -0.030608  1.000000

Estudiando las columnas en conjunto podemos concluir que la columna 'name', no tendra practicamente interés para el modelo, ya que no oferce informacion relevante y a simplpe vista no parece tener correlacion entre el precio y el numero de visitas.
Además, sabiendo el mercado del que se trata, es decir inmuebles turísticos de Airbnb, podemos concluír que a la hora de hacer una búsqueda de alojamiento, el precio y la situación son las principales variables, y que el nombre apenas tiene interés ni llamada al 'click', y que en todo caso seria da descripcion detallada y sobre todo las fotografía lo que podría inclinar la balanza de la demanda.

In [32]:
data[['number_of_reviews','reviews_per_month', 'last_review']].isnull().mean()*100

number_of_reviews     0.000000
reviews_per_month    20.558339
last_review          20.558339
price                 0.000000
dtype: float64

La correlación entre 'number_of_reviews' y 'reviews_per_month' es positiva moderada (0.5499).
Puesto un mayor número total de reseñas también tengan una tasa de reseñas mensuales más alta, tienen una tasa del 20%, que es un nº considerable, y la información que ofrece no es relevante, ya que no tiene apenas correlación con el precio (variable clave), eliminaremos tambien estas columnas.

In [33]:
data.drop(['id', 'name', 'host_id', 'host_name', 'number_of_reviews','last_review', 'reviews_per_month'], axis=1, inplace=True)
data.head()

neighbourhood_group neighbourhood  latitude  longitude        room_type  \
0            Brooklyn    Kensington  40.64749  -73.97237     Private room   
1           Manhattan       Midtown  40.75362  -73.98377  Entire home/apt   
2           Manhattan        Harlem  40.80902  -73.94190     Private room   
3            Brooklyn  Clinton Hill  40.68514  -73.95976  Entire home/apt   
4           Manhattan   East Harlem  40.79851  -73.94399  Entire home/apt   

   price  minimum_nights  calculated_host_listings_count  availability_365  
0    149               1                               6               365  
1    225               1                               2               355  
2    150               3                               1               365  
3     89               1                               1               194  
4     80              10                               1                 0

In [34]:
# Buscamos duplicados
data.duplicated().sum()

0

In [35]:
# Buscamos valores nulos en todo el conjunto
data.isnull().any().any()

False

In [36]:
data.describe().T

count        mean         std       min  \
latitude                        48895.0   40.728949    0.054530  40.49979   
longitude                       48895.0  -73.952170    0.046157 -74.24442   
price                           48895.0  152.720687  240.154170   0.00000   
minimum_nights                  48895.0    7.029962   20.510550   1.00000   
calculated_host_listings_count  48895.0    7.143982   32.952519   1.00000   
availability_365                48895.0  112.781327  131.622289   0.00000   

                                     25%        50%         75%          max  
latitude                        40.69010   40.72307   40.763115     40.91306  
longitude                      -73.98307  -73.95568  -73.936275    -73.71299  
price                           69.00000  106.00000  175.000000  10000.00000  
minimum_nights                   1.00000    3.00000    5.000000   1250.00000  
calculated_host_listings_count   1.00000    1.00000    2.000000    327.00000  
availability_365                 0.00000   45.00000  227.000000    365.00000

 - Latitude y Longitude:
La media de la latitud es aproximadamente 40.73 y la media de la longitud es aproximadamente -73.95.
La desviación estándar para ambas características es relativamente baja, lo que sugiere que la mayoría de las ubicaciones se encuentran cerca de la media.
Los valores mínimos y máximos de latitud y longitud indican que todos los datos se encuentran dentro de un rango razonable para la ciudad de Nueva York.

 - Precio (Price):
El precio medio es de alrededor de $152, con una desviación estándar bastante alta de $240.
El rango de precios varía desde $0 hasta $10,000, lo que indica una amplia gama de precios en los alojamientos.
La mayoría de los precios se concentran por debajo de los $175, como lo demuestra el percentil 75.

 - Número mínimo de noches (Minimum Nights):
El promedio de noches mínimas es de aproximadamente 7, pero la desviación estándar es alta (aproximadamente 20.51), lo que indica una gran variabilidad en este aspecto.
El valor máximo de 1250 noches parece ser un valor atípico y podría ser el resultado de errores o entradas incorrectas en los datos.

 - Número de listados calculados por el anfitrión (Calculated Host Listings Count):
El promedio de listados calculados por el anfitrión es de alrededor de 7, pero la desviación estándar es muy alta, lo que sugiere una gran variabilidad en el número de listados por anfitrión.
El valor máximo de 327 parece ser un valor atípico y puede ser el resultado de anfitriones con múltiples propiedades en la plataforma.
Sin embargo, tal y como observamnos anetriormente, 'host_name', de entre todos los anuncios listados (48895), 48874 son distintos, es decir, solo hay 21 inmuebles que podrían tener el mismo propietario, con lo cual este valor máximo sigue pareciendo desproporcionado.

 - Disponibilidad durante el año (Availability 365):
La disponibilidad media durante el año es de aproximadamente 112 días, pero la desviación estándar es bastante alta, lo que indica una gran variabilidad en la disponibilidad.
El percentil 75 muestra que la mayoría de los alojamientos tienen una disponibilidad de menos de 227 días al año.

In [37]:
data[data['price'] == 0]
print((data['price'] == 0).sum())
data.drop(data[data['price'] == 0].index, axis=0, inplace=True)

# Observamos que hay 11 inmuebles a precio 0. Esto debe ser un error y es proporcionalmente nimio, limpiamos estas filas.

11


In [38]:
print((data['price'] > 500).sum())
print(f"{((data['price'] > 500).sum()) / len(data['price']) * 100}%")
data.drop(data[data['price'] > 500].index, axis=0, inplace=True)

# Lo mismo con los inmuebles de más de 500$, que suponen tan solo el 2% del conjunto, y sería mas efectivo observaros en un modelo 'de lujo'.

1044
2.135668112265772%


In [39]:
# Es posible que al filtrar y eliminar algunas filas del DataFrame original usando el método dropna() con inplace=True, el índice del DataFrame no se haya ajustado correctamente,
# lo que puede llevar a que el DataFrame resultante tenga índices adicionales que no corresponden a las filas eliminadas.

data.reset_index(drop=True, inplace=True)  # Restablece el índice del DataFrame después de eliminar filas no deseadas


In [40]:
fig, axis = plt.subplots(2, 2, figsize = (10, 7))

# histograma múltiple
sns.histplot(ax = axis[0, 0], data = data, x = "neighbourhood_group").set(ylabel=None)
sns.histplot(ax = axis[0, 1], data = data, x = "neighbourhood").set(ylabel = None)
sns.histplot(ax = axis[1, 0], data = data, x = "room_type").set(ylabel = None)
sns.histplot(ax = axis[1, 1], data = data, x = "price").set(ylabel = None)

# Rotar nombres del eje x en el subplot neighbourhood

#axis[0, 1].set_xticklabels(axis[0, 1].get_xticklabels(), rotation=45)

# Ajustar el layout
plt.tight_layout()

plt.show()

NameError: name 'plt' is not defined